main/main.cpp

```diff
@@ -0,0 +1,191 @@
+using namespace ::nl::Weave;
+using namespace ::nl::Weave::DeviceLayer;
+
+#define UNUSED_PARAMETER(a) (a = a)
+
+volatile int apperror_cnt;
+
+// ================================================================================
+// App Error
+//=================================================================================
+void appError(int err)
+{
+    EFR32_LOG("!!!!!!!!!!!! App Critical Error: %d !!!!!!!!!!!", err);
+    portDISABLE_INTERRUPTS();
+    while (1)
+        ;
+}
+
+// ================================================================================
+// FreeRTOS Callbacks
+// ================================================================================
+extern "C" void vApplicationIdleHook(void)
+{
+    // FreeRTOS Idle callback
+
+    // Check Weave Config nvm3 and repack flash if necessary.
+    Internal::EFR32Config::RepackNvm3Flash();
+}
+
+// ================================================================================
+// Main Code
+// ================================================================================
+int main(void)
+{
+    int ret = WEAVE_ERROR_MAX;
+
+    otSysInit(0, NULL);
+
+    // Initialize mbedtls threading support on EFR32
+    THREADING_setup();
+
+    BSP_LedsInit();
+    BSP_LedSet(0);
+    BSP_LedSet(1);
+
+#if EFR32_LOG_ENABLED
+    if (efr32LogInit() != 0)
+    {
+        appError(ret);
+    }
+#endif
+
+    EFR32_LOG("==================================================");
+    EFR32_LOG("openweave-efr32-lock-example starting");
+    EFR32_LOG("==================================================");
+
+    // Configure mbedTLS to use FreeRTOS-based mutexes.  This ensures that mbedTLS can be used
+    // simultaneously from multiple FreeRTOS tasks (e.g. OpenThread, OpenWeave and the application).
+    // TODO freertos_mbedtls_mutex_init();
+
+    EFR32_LOG("Initializing Weave stack");
+    ret = PlatformMgr().InitWeaveStack();
+    if (ret != WEAVE_NO_ERROR)
+    {
+        EFR32_LOG("PlatformMgr().InitWeaveStack() failed");
+        appError(ret);
+    }
+
+    EFR32_LOG("Initializing OpenThread stack");
+    ret = ThreadStackMgr().InitThreadStack();
+    if (ret != WEAVE_NO_ERROR)
+    {
+        EFR32_LOG("ThreadStackMgr().InitThreadStack() failed");
+        appError(ret);
+    }
+
+    // Reconfigure mbedTLS to use regular calloc and free.
+    //
+    // By default, OpenThread configures mbedTLS to use its private heap at initialization time.  However,
+    // the OpenThread heap is not thread-safe, effectively preventing other threads from using mbedTLS
+    // functions.
+    //
+    // Note that this presumes that the system heap is itself thread-safe.  On newlib-based systems
+    // this requires a proper implementation of __malloc_lock()/__malloc_unlock() for the applicable
+    // RTOS.  It also requires the heap to be provisioned with enough storage to accommodate OpenThread's
+    // needs.
+    //
+    // TODO mbedtls_platform_set_calloc_free(calloc, free);
```

Comment:
@marvensilabs Can you comment on how mbedtls memory allocation works on the EFR32 platforms?  In particular, is it thread thread-safe by default, and does it use the same heap as malloc/free()?

Reply:
If mbedtls's use of malloc isn't an issue, then I suggest removing this comment.

Reply:
By default mbedtls should use Newlib's calloc and free which are not thread-safe by default in our SDK but I see that the __malloc_lock / _malloc_unlock functions in main/support/FreeRTOSNewlibLockSupport.c are being called to make it thread-safe.

However I see that we are calling mbedtls_platform_set_calloc_free to set the allocator functions to CAlloc and Free in OpenThread's mbedtls.cpp.  It looks like we are using the OpenThread CAlloc implementation in openthread/src/core/utils/heap.cpp which appears to manage it's own pool of memroy without calling malloc and free.

Reply:
Since mbedtls will be called from multiple tasks (thread, weave and potentially the application) it must be configured to use a thread-safe allocator, which the OpenThread allocator is not.

The code that was commented out is one way to achieve this.  However looking at the OpenThread code, it seems if you define OPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS_MANAGEMENT to 0, OpenThread will *not* override the default mbedtls allocator, leaving it to call standard calloc() and free().

Thus I suggest adding this define to OpenThreadConfig.h and the double checking that the right functions are being called.

Reply:
I could not find OPENTHREAD_CONFIG_ENABLE_BUILTIN_MBEDTLS_MANAGEMENT in the commit of OpenThread I'm using but could find OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE which seems to cover all allocations made by OpenThread and not just MBEDTLS.  I've added this define to the efr32-openthread.mk and reinstated the call to mbedtls_platform_set_calloc_free into main.cpp.

Reply:
Apparently I was looking at an older version of OpenThread. 

I see that rather than calling mbedtls_platform_set_calloc_free() you're calling otHeapSetCAllocFree.  That should be fine, however remember to size your malloc heap to accommodate the increased demand from OpenThread.

Also, I would prefer that OPENTHREAD_CONFIG_HEAP_EXTERNAL_ENABLE were set in OpenThreadConfig.h within the example app, rather than in efr32-openthread.mk in openweave-core.  That way individual projects can reconfigure the heaps the way they see fit.

Let me know when your testing is complete.